src/explorer/CosmosDBTreeItem.ts

```diff
@@ -159,10 +171,48 @@ export class CosmosDBTreeItem extends AzureParentTreeItem<ISiteTreeRoot> {
         return result;
     }
 
+    private detectPostgresConnections(appSettings: { [propertyName: string]: string }): IDetectedConnection[] {
+        const connectionStringPrefix = 'postgres://';
+        const portDefault = '5432';
+
+        const result: IDetectedConnection[] = [];
+        const regexp = new RegExp(`(.+)${this._pgHostSuffix}`, 'i');
+        for (const key of Object.keys(appSettings)) {
+            const match = key && key.match(regexp);
+            if (match) {
+                const prefix = match[1];
+                const hostKey = prefix + this._pgHostSuffix;
+                const dbNameKey = prefix + this._pgDbNameSuffix;
+                const userKey = prefix + this._pgUserSuffix;
+                const passKey = prefix + this._pgPassSuffix;
+                const portKey = prefix + this._pgPortSuffix;
+                const host: string | undefined = appSettings[hostKey];
+                const dbName: string | undefined = appSettings[dbNameKey];
+                const username: string | undefined = appSettings[userKey];
+                const password: string | undefined = appSettings[passKey];
+                const port: string | undefined = appSettings[portKey];
+
+                if (host) {
+                    const keys: string[] = [hostKey];
+                    keys.push(userKey, passKey, portKey, dbNameKey);
+                    let connectionString: string = connectionStringPrefix;
+                    if (username && password) {
+                        connectionString += `${username}:${password}@`;
+                    }
+                    connectionString += `${host}:${port ? port : portDefault}`;
+                    if (dbName) {
+                        connectionString += `/${dbName}`;
```

Comment:
Technically this should do all the same encoding stuff you're doing in the cosmos repo with encodeUriComponent. However, there's another option...

The connectionString is only used as a part of `TreeItemQuery` in `findTreeItem`. And the only reason we did it that way is because Mongo is the first one we added support for and it saves the app setting as an entire connection string. Since postgres splits up the connection string into multiple app settings, I think it makes total sense to add another option on the query like this:
```
export interface TreeItemQuery {
    /**
     * An account or database connection string
     */
    connectionString?: string;

    postgresData?: {
        hostName?: string;
        // etc.
    }
}
```

That way we don't have to construct the connection string here (and deal with all the encoding issues, etc.) just to have it deconstructed in our own code in the cosmos repo

Reply:
I agree with you on adding encoding, however I would prefer to keep it with connectionString for this release as the changes here would require quite a few changes in the api code in Databases as well.

Reply:
Yeah docDbData could do the same thing, but it's been using connectionString for a while, so I don't mind keeping it that way.

On the other hand, neither PR for the postgres connections api (in this repo or Cosmos) has been merged, so it feels a bit premature to delay this kind of change. The suggestion is meant to make your code _less_ complex (aka easier to review/test) because you don't have to encode/decode/parse the connection string on both sides. I also err on the side of putting the work in up-front when dealing with an API between two extensions because of all the versioning complexity involved.